### ISSUE TITLE
DiffusionDet: Diffusion Model for Object Detection

### ISSUE THREAD

**Issue by kabupen** [kabupen/papers#6]
# 

# Approach

### object detection

物体検出は以下の情報を予測対象とする

- $x$ ：入力画像
- $b = (c_x^i, c_y^i, w^i, h^i)$ 
- $c$ ：カテゴリクラス

### Diffusion model

拡散過程は

$$
q(z_t|z_0) = N(z_t|\sqrt{\bar{\alpha_t}} z_0, (1-\bar{\alpha_t}) I ))
$$

で表され、学習時には逆拡散過程の出力である $f_\theta(z_t, t)$ が入力元画像である $z_0$ を再現できるように学習を進める。

$$
L = \frac{1}{2} || f_\theta(z_t, t) - z_0 ||^2
$$

の L2 loss が最適化の対象であった。

本研究では $z_0 = b$ であり、$b \in R^{N\times 4}$ （4次元ベクトルが $N$ 個の bbox 分）である。$z_t$ は "noisy boxes" であり、 $z_t$ から真の bbox である $z_0$ を予測する。

## Architecture

ここでは通常の拡散モデルと異なり $f(z_t, t, x)$ を計算するため（ノイズbbox、ステップ数、実写真が入力）、推論時には逆拡散過程ごとに raw image を処理する必要が生じてしまう。そこで encoder と detection decorder と2段階に分けて、raw image を特徴量ベクトルに変換してから使用することにしている。

![image](https://github.com/kabupen/papers/assets/19812756/ebca7cdc-e876-4750-a0f6-5105c87d244c)

- 明示はないが、出力の class はクラス数次元のベクトル？になっていて、それぞれが確率としての意味をもっているのだと思われる。
   - $N$ 個の出力が出てきたときに、"犬" ラベルのものを抽出する（かつ確率値の高いもの）ような処理が必要となる 

### Sparse R-CNN

![image](https://github.com/kabupen/papers/assets/19812756/fe3769ce-3bf1-498a-a2b7-43a1460e689f)

### 比較

- モデル構造は基本的には似ている（じゃっかんの思想が違う）
- 推論時にはDiffusionDet は完全ランダムな bbox から出発するが、Sparse R-CNN は固定の learned bbox から出発する
    - つまり事前知識として「ここに物体があるっぽい」 「ここを提案します」というようなものが必要ないということ
- Sparse R-CNN では proposal bbox で画像をクロップして特徴量を計算しているが、DiffusionDet では画像を一括で embedding しているので、都度特徴量計算がいらない

### 利点？

- 推論時には適当なノイズの bbox を用意して、それを逆拡散過程に入れることで予測bbox が出る。このときに用意するノイズbboxの個数 $N_{eval}$ は動的に変更することができ、$N_{train}$ と同じではなくていい
    - sparse/crowded な環境どちらにも対応できる 
- head を使い回すことでより性能を求めることができる（ステップ数を増やせばいいとかかな？）

### Ground truth boxed padding

画像の中のbbox数はそれぞれ異なっているため、$N_{train}$ 個数の bbox が画像内に存在するようにまずは "padding" を実施する（画像処理の padding というよりは単に $N^\prime$ 個の bbox の個数が $N_{train}$ になるように埋める、くらいの意味）。このランダムな bbox の作り方は色々試した結果、ガウス分布からのランダムサンプリングが良かったとのこと。

### Box corruption

ground truth bboxes にノイズを加えて、bbox を"破壊"する。

### Training Loss

detection decoder は $N_{train}$ 個の bbox を入力として受け取り、$N_{train}$ 個のカテゴリと座標とを予測する。Sparse R-CNN のものを拝借している感じで

![image](https://github.com/kabupen/papers/assets/19812756/9af61360-1ad1-4d28-a720-990333e61157)

クラス分類の損失、その他の損失を合わせたものを予測損失としている。

### Sampling steps

前時刻からの予測 bbox （or 初期値ならランダム bbox）を detection decoder に入力して、クラスと座標値を予測する。初期値はランダムに設定でき、かつ何個の bbox を用意するかも任意である。

### Box renewal 

予測した bbox は desired/undesired に分類される。undesied bbox（おそらくbboxではないもの）を次のステップに入力するのはよくないので、random bbox に置換する方法を bbox renewal として導入した。つまり、denoising している最中に「これは意味のないbboxだ」と判定された場合には次の denoising を行わずに、初期化して denoising にかけるということ。
